node/src/components/diagnostics_port.rs

```diff
@@ -0,0 +1,247 @@
+            }
+        }
+    }
+
+    // This is not thread-safe, as it will set the umask for the entire process, but we assume that
+    // initalization happens "sufficiently single-threaded".
+    let umask_guard = umask::temp_umask(socket_umask);
+    let listener = UnixListener::bind(socket_path)?;
+    drop(umask_guard);
+
+    debug!(local_addr=%ShowUnixAddr(&listener.local_addr()?), "diagnostics port listening");
+
+    Ok(listener)
+}
+
+/// Diagnostics port event.
+#[derive(Debug, Serialize)]
+pub(crate) struct Event;
+
+/// A diagnostics port initialization error.
+#[derive(Debug, Error)]
+pub(crate) enum Error {
+    /// Error setting up the diagnostics port's unix socket listener.
+    #[error("could not setup diagnostics port listener")]
+    SetupListener(#[from] io::Error),
+}
+
+impl Display for Event {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str("diagnostics port event")
+    }
+}
+
+impl<REv> Component<REv> for DiagnosticsPort {
+    type Event = Event;
+
+    type ConstructionError = Error;
+
+    fn handle_event(
+        &mut self,
+        _effect_builder: EffectBuilder<REv>,
+        _rng: &mut NodeRng,
+        _event: Event,
+    ) -> Effects<Event> {
+        // No events are processed in the component, as all requests are handled per-client in
+        // tasks.
+        Effects::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        fs,
+        os::unix::prelude::{FileTypeExt, PermissionsExt},
+    };
+
+    use tokio::{
+        io::{AsyncReadExt, AsyncWriteExt},
+        net::UnixStream,
+    };
+
+    use super::setup_listener;
+
+    #[tokio::test]
+    async fn setup_listener_creates_listener() {
+        let tmpdir = tempfile::tempdir().expect("could not create tempdir");
+        let socket_path = tmpdir.path().join("test.socket");
+
+        // We give it a strict umask to check.
+        let listener = setup_listener(&socket_path, 0o077).expect("could not setup listener");
+
+        let meta = fs::metadata(&socket_path).expect("could not get meatadata");
+        // With the given umask, world and group permissions should be 0.
+        assert_eq!(meta.permissions().mode() & 0o077, 0);
+
+        // Attempt to connect.
+        tokio::spawn(async move {
+            let mut stream = UnixStream::connect(socket_path)
+                .await
+                .expect("could not connect to listener");
+            stream
+                .write_all(b"hello")
+                .await
+                .expect("could not write to listener");
+        });
+
+        let (mut stream, _socket_addr) = listener
+            .accept()
+            .await
+            .expect("could not accept connection");
+
+        let mut buffer = Vec::new();
+        stream
+            .read_to_end(&mut buffer)
+            .await
+            .expect("failed to read to end");
+        assert_eq!(b"hello", buffer.as_slice());
```

Comment:
Nitpick but `hello` could be extracted to a variable and shared between the writing task and here.